File: src/Common/Packet.cpp
/*
Project Faolan a Simple and Free Server Emulator for Age of Conan.
Copyright (C) 2009, 2010 The Project Faolan Team

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include "Packet.h"

Packet::Packet(PacketBuffer* pBuffer)
{
	packetBuffer = pBuffer;
	length = packetBuffer->read<uint32>();
	crc32 = packetBuffer->read<uint32>();
	headersize = packetBuffer->read<uint32>(); //size of header
	sender = packetBuffer->read<uint8>();
	slength = packetBuffer->read<uint8>();
	if(slength<=0)
		printf("Error at senderlength\n");
	senderInt = new PacketBuffer(packetBuffer->readArray(slength), slength);
	receiver = packetBuffer->read<uint8>();
	rlength = packetBuffer->read<uint8>();
	if(rlength<=0)
		printf("Error at receiverlength\n");
	receiverInt = new PacketBuffer(packetBuffer->readArray(rlength), rlength);

	uint32 checkHeaderLength = headersize - (4 + slength + rlength);

	if(checkHeaderLength == 2)
	{
		opcode = packetBuffer->read<uint16>();
	}
	else if(checkHeaderLength == 4)
	{
		opcode = packetBuffer->read<uint32>();
	}
	else if(checkHeaderLength > 4)
	{
		headerData = new PacketBuffer(packetBuffer->readArray((checkHeaderLength - 4)), (checkHeaderLength - 4));
		opcode = packetBuffer->read<uint32>();
	}
	else if(checkHeaderLength == 3)
	{
		opcode = packetBuffer->read<uint16>();
		headerData = new PacketBuffer(packetBuffer->readArray((checkHeaderLength - 2)), (checkHeaderLength - 2));
	}
	else
	{
		printf("ERROR at Packet.cpp (Headerstruct)\nPacketLength: 0x%08x\nheaderSize: 0x%08x\n", length, headersize);
	}

	uint32 dataLength = length - (headersize + 8);
	if(dataLength > 0)
		data = new PacketBuffer(packetBuffer->readArray(dataLength), dataLength);
	else
	{
		data = new PacketBuffer(0);
	}
}

uint32 Packet::HeaderLength()
{
	return (headersize + 8);
}

void Packet::drop()
{
	delete senderInt;
	delete receiverInt;
	delete headerData;
	delete data;
}

File: src/LibFaolan/Network/Packet.cs
using System;
using Hik.Communication.Scs.Communication.Messages;
using LibFaolan.Extentions;

namespace LibFaolan.Network
{
    public sealed class Packet : IScsMessage
    {
        /////////
        /*
            UInt32 length;
            UInt32 crc32;
            UInt32 headersize;
            byte sender;
            byte slength;
            byte[] senderInt;
            byte receiver;
            byte rlength;
            byte[] receiverInt;
            UInt32 opcode;
            byte[] headerData;
            byte[] data; 
        */

        private readonly ConanStream originalStream;
        public UInt32 Crc32;
        public ConanStream Data;
        public ConanStream HeaderData;
        public UInt32 Headersize;
        public UInt32 Length;
        public UInt32 Opcode;
        public byte Receiver;
        public byte[] ReceiverInt;
        public byte Sender;
        public byte[] SenderInt;

        public Packet(ConanStream stream)
        {
            originalStream = stream;
            stream.Position = 0;

            Length = stream.ReadUInt32();

            if (Length == 0 || stream.Length < Length)
            {
                //throw new Exception("stream.Length < length");
                Length = UInt32.MaxValue;
                return;
            }

            Crc32 = stream.ReadUInt32();
            Headersize = stream.ReadUInt32();
            if (Headersize > 0xffff) // this should do..
            {
                //throw new Exception("Headersize > 0xffff");
                Length = UInt32.MaxValue;
                return;
            }

            Sender = stream.ReadByte();
            SenderInt = stream.ReadArrayPrependLengthByte();

            Receiver = stream.ReadByte();
            ReceiverInt = stream.ReadArrayPrependLengthByte();

            Opcode = stream.ReadUInt16();

            var checkHeaderLength = Headersize - (UInt32) (sizeof (UInt32) + SenderInt.Length + ReceiverInt.Length);
            if (checkHeaderLength > 0)
                HeaderData = new ConanStream(stream.ReadArray(checkHeaderLength - sizeof (byte)*2));

            var dataLength = Length - (Headersize + sizeof (UInt32)*2);
            if (dataLength > 0)
                Data = new ConanStream(stream.ReadArray(dataLength));
        }

        public UInt32 HeaderLength => (Headersize + sizeof (UInt32)*2); // + length & crc32
        public string MessageId => null;

        public string RepliedMessageId
        {
            get { return null; }
            set { }
        }

        public override string ToString()
        {
            return string.Format("Opcode: {0} Length: {1}", Opcode.ToHex(), originalStream.Length);
        }
    }
}